Add workflow for SARS-CoV-2 variant analysis

pvanheus opened the pull request:
This is a workflow for Illumina SARS-CoV-2 variant analysis with variant calling and consensus construction using iVar and lineage/clade assignment using pangolin and nextclade.

bgruening commented:
@pvanheus can you add some more metadata please. E.g. a changelog file and dockstore.yml?

![image](https://user-images.githubusercontent.com/469983/122684841-4094d880-d208-11eb-841f-7a2586e3cdf0.png)

bwlang commented:
@pvanheus I'm excited about this one... let me know how I can help.

pvanheus commented:
> @pvanheus I'm excited about this one... let me know how I can help.

Thanks for your interest! I'm also working on [this tutorial](https://github.com/galaxyproject/training-material/pull/2633).

As [you](https://github.com/andersen-lab/ivar/issues/90) and others have noted, iVar is not without its issues. Expanding the set of test data to illustrate where problems might be would be useful. There have been a variety of approaches to addressing these, e.g. [this workflow](https://github.com/galaxyproject/iwc/tree/main/workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-variant-calling) from @wm75 and [this from covid-19-signal](https://github.com/jaleezyy/covid-19-signal/blob/master/Snakefile#L630) that uses Freebayes in addition to iVar. I think its worth getting this workflow in its current form out there so long though since this approach is in wide use worldwide.

pvanheus commented:
Workflow testing is currently failing because `nextflow` is not available on usegalaxy.org. Waiting on pull requests from @mvdbeek to be merged to resolve this.

mvdbeek commented on workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis.ga:
```diff
@@ -0,0 +1,1215 @@
+        },
+        "19": {
+            "annotation": "",
+            "content_id": "toolshed.g2.bx.psu.edu/repos/iuc/nextclade/nextclade/0.14.3+galaxy1",
```
The latest version seems to be 0.14.4+galaxy0, any reason not to use that one ?
```suggestion
            "content_id": "toolshed.g2.bx.psu.edu/repos/iuc/nextclade/nextclade/0.14.4+galaxy0",
```

mvdbeek commented on workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis.ga:
```diff
@@ -0,0 +1,1215 @@
+            },
+            "tool_id": "toolshed.g2.bx.psu.edu/repos/iuc/nextclade/nextclade/0.14.3+galaxy1",
+            "tool_shed_repository": {
+                "changeset_revision": "6cd62b6d2b5c",
```
```suggestion
                "changeset_revision": "3b75d5285aff",
```

mvdbeek commented on workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis.ga:
```diff
@@ -0,0 +1,1215 @@
+                "tool_shed": "toolshed.g2.bx.psu.edu"
+            },
+            "tool_state": "{\"adv\": {\"advanced_options\": \"no\", \"__current_case__\": 1}, \"include_header\": \"false\", \"input_fasta\": {\"__class__\": \"ConnectedValue\"}, \"outputs\": [\"report_tsv\"], \"__page__\": null, \"__rerun_remap_job_id__\": null}",
+            "tool_version": "0.14.3+galaxy1",
```
```suggestion
            "tool_version": "0.14.4+galaxy0",
```

mvdbeek commented on workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis.ga:
```diff
@@ -0,0 +1,1215 @@
+                    "output_name": "report_tsv"
+                }
+            },
+            "tool_id": "toolshed.g2.bx.psu.edu/repos/iuc/nextclade/nextclade/0.14.3+galaxy1",
```
```suggestion
            "tool_id": "toolshed.g2.bx.psu.edu/repos/iuc/nextclade/nextclade/0.14.4+galaxy0",
```

mvdbeek commented on workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis.ga:
```diff
@@ -0,0 +1,1215 @@
+        },
+        "16": {
+            "annotation": "",
+            "content_id": "toolshed.g2.bx.psu.edu/repos/iuc/multiqc/multiqc/1.8+galaxy1",
```
```suggestion
            "content_id": "toolshed.g2.bx.psu.edu/repos/iuc/multiqc/multiqc/1.9+galaxy1",
```

mvdbeek commented on workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis.ga:
```diff
@@ -0,0 +1,1215 @@
+                    "output_name": "html_report"
+                }
+            },
+            "tool_id": "toolshed.g2.bx.psu.edu/repos/iuc/multiqc/multiqc/1.8+galaxy1",
```
```suggestion
            "tool_id": "toolshed.g2.bx.psu.edu/repos/iuc/multiqc/multiqc/1.9+galaxy1",
```

mvdbeek commented on workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis.ga:
```diff
@@ -0,0 +1,1215 @@
+            },
+            "tool_id": "toolshed.g2.bx.psu.edu/repos/iuc/multiqc/multiqc/1.8+galaxy1",
+            "tool_shed_repository": {
+                "changeset_revision": "5e33b465d8d5",
```
```suggestion
                "changeset_revision": "75c93c70d094",
```

mvdbeek commented on workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis.ga:
```diff
@@ -0,0 +1,1215 @@
+                "tool_shed": "toolshed.g2.bx.psu.edu"
+            },
+            "tool_state": "{\"comment\": \"\", \"flat\": \"false\", \"results\": [{\"__index__\": 0, \"software_cond\": {\"software\": \"fastp\", \"__current_case__\": 7, \"input\": {\"__class__\": \"RuntimeValue\"}}}, {\"__index__\": 1, \"software_cond\": {\"software\": \"samtools\", \"__current_case__\": 23, \"output\": [{\"__index__\": 0, \"type\": {\"type\": \"stats\", \"__current_case__\": 0 ...
+            "tool_version": "1.8+galaxy1",
```
```suggestion
            "tool_version": "1.9+galaxy1",
```

mvdbeek commented on workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis.ga:
```diff
@@ -0,0 +1,1215 @@
+                "owner": "iuc",
+                "tool_shed": "toolshed.g2.bx.psu.edu"
+            },
+            "tool_state": "{\"comment\": \"\", \"flat\": \"false\", \"results\": [{\"__index__\": 0, \"software_cond\": {\"software\": \"fastp\", \"__current_case__\": 7, \"input\": {\"__class__\": \"RuntimeValue\"}}}, {\"__index__\": 1, \"software_cond\": {\"software\": \"samtools\", \"__current_case__\": 23, \"output\": [{\"__index__\": 0, \"type\": {\"type\": \"stats\", \"__current_case__\": 0 ...
```
```suggestion
            "tool_state": "{\"comment\": \"\", \"export\": \"false\", \"flat\": \"false\", \"results\": [{\"__index__\": 0, \"software_cond\": {\"software\": \"fastp\", \"__current_case__\": 7, \"input\": {\"__class__\": \"RuntimeValue\"}}}, {\"__index__\": 1, \"software_cond\": {\"software\": \"samtools\", \"__current_case__\": 23, \"output\": [{\"__index__\": 0, \"type\": {\"type\": \"stats\", \"__current_case__\": 0, \"input\": {\"__class__\": \"RuntimeValue\"}}}]}}, {\"__index__\": 2, \"software_cond\": {\"software\": \"qualimap\", \"__current_case__\": 19, \"input\": {\"__class__\": \"RuntimeValue\"}}}], \"saveLog\": \"false\", \"title\": \"\", \"__page__\": null, \"__rerun_remap_job_id__\": null}",
```

mvdbeek commented:
@pvanheus it's green now, but I've suggested updates of nextclade and multiqc. Let me know if you want to merge as is and then update, or update right away

drosofff commented on workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis.ga:
```diff
@@ -0,0 +1,1215 @@
+        },
+        "10": {
+            "annotation": "",
+            "content_id": "toolshed.g2.bx.psu.edu/repos/iuc/qualimap_bamqc/qualimap_bamqc/2.2.2d+galaxy1",
```
Hi @pvanheus there is also a newer version for this tool: qualimap_bamqc/2.2.2d+galaxy3

wm75 commented on workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis.ga:
```diff
@@ -0,0 +1,1215 @@
+        },
+        "7": {
+            "annotation": "",
+            "content_id": "toolshed.g2.bx.psu.edu/repos/devteam/samtool_filter2/samtool_filter2/1.8+galaxy1",
```
Can you use samtools_view - toolshed.g2.bx.psu.edu/repos/iuc/samtools_view/samtools_view/1.9+galaxy3 - instead of this outdated tool please?

wm75 commented on workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis.ga:
```diff
@@ -0,0 +1,1215 @@
+                "owner": "iuc",
+                "tool_shed": "toolshed.g2.bx.psu.edu"
+            },
+            "tool_state": "{\"amplicons\": {\"filter_by\": \"\", \"__current_case__\": 1}, \"inc_primers\": \"true\", \"input_bam\": {\"__class__\": \"ConnectedValue\"}, \"min_len\": \"30\", \"min_qual\": \"20\", \"primer\": {\"source\": \"history\", \"__current_case__\": 0, \"input_bed\": {\"__class__\": \"ConnectedValue\"}}, \"primer_pos_wiggle\": \"0\", \"window_width\": \"4\", \"__page__\": n ...
```
You may want to consider "filter_by": "yes" here to make use of ivar trim's new `-f` option, which makes use of knowledge about expected amplicons to discard reads that do not fall within amplicon boundaries.

The downside is that you'll require one more input dataset, but it seems to be worth it -> https://github.com/andersen-lab/ivar/issues/75

wm75 replied:
This is one of the improvements in v0.3 of (https://github.com/galaxyproject/iwc/tree/main/workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-variant-calling)

wm75 commented on workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis.ga:
```diff
@@ -0,0 +1,1215 @@
+        },
+        "14": {
+            "annotation": "",
+            "content_id": "toolshed.g2.bx.psu.edu/repos/bgruening/text_processing/tp_sed_tool/1.1.1",
```
This step seems *very* questionable!

Just renaming the reference to have it recognized by SnpEff can mask input data issues. What if the reference used by the user isn't identical to NC_045512.2?

wm75 replied:
My assumption is you do this to keep both ivar trim and SnpEff happy?
If that's the case you can just omit this step and directly use NC_045512.2 as input. ivar trim will change the primer bed file ref name on the fly for you.

pvanheus replied:
The renaming is because MN908947.3 is often used as input and yes it does have a risk like you outlined.

wm75 replied:
Ok, then what about this compromise: You replace specifically only MN908947.3 with NC_045512.2, but do this up front on the user-provided fasta file?

Those two are base identical, correct?

bgruening replied:
Should we not explicitly state which input we assume, which reference genome, and not hack around potential user assumptions. This is a general question for all workflows under IWC.

wm75 replied:
In general, I agree (need to check our existing WFs 😊 ). In the specific case though MN908947.3 is the ID of the record the refseq NC_045512.2 was generated from and the two are identical and used kind of interchangeably (the ARTIC networks primer scheme files, e.g., use MN908947.3 instead of the refseq ID). They are identical (I just reconfirmed) so there's a point to making the WF convenient to use.
I don't think I will add such an extra step to the covid19.galaxyproject.org WFs, but this one here really emphasizes end-user friendliness so maybe it's an appropriate thing to do here.
As I said, I would like to see the replacement done as the very first thing in the WF though. No changing of IDs half-way through the history.

wm75 commented on workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis.ga:
```diff
@@ -0,0 +1,1215 @@
+        },
+        "18": {
+            "annotation": "",
+            "content_id": "toolshed.g2.bx.psu.edu/repos/iuc/pangolin/pangolin/3.0.3+galaxy0",
```
This one absolutely needs refactoring.

Running Pangolin on a collection of Fastas means independent jobs will all download the latest database from the web!
Please concatenate the Fastas, not the pangolin output.

Probably the same works for nextclade? Which would even save a step in the WF then cause you'd have to concatenate only once upstream?

wm75 commented on workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis.ga:
```diff
@@ -0,0 +1,1215 @@
+            "post_job_actions": {
+                "RenameDatasetActionoutput": {
+                    "action_arguments": {
+                        "newname": "#{infile}"
```
Can you find a better name here?

What this does is it will create datasets named "ivar consensus on data xxx Consensus" just like the previous step (i.e. the name will be misleading both in terms of the tool and the input dataset #). I know you're hiding the previous step's output collection, but if you ever need to debug one of the resulting histories, you'll regret this choice.

pvanheus replied:
In my experience, this renames the outputs to have the name of the collection element - i.e. if the input is called `sample1` the output is `sample1` in the collection. The naming of the collection as a whole remains `"iv.ar consensus on data xxx Consensus`.

wm75 replied:
The renaming applies to the collection and its datasets. It's the collection elements' element identifers that are passed through like you say. In other words when you're viewing the datasets in the context of the collection you will see the names you're expecting, but when you view them as individual datasets (e.g. when you unhide hidden elements) or try to edit their names, you will see the "newname" set here.
Anyway, that newname is misleading for the collection and its elements.

wm75 commented:
As a general comment, you may want to work a bit more on the dataset/collection names of the workflow outputs.

![Screenshot from 2021-06-21 18-12-55](https://user-images.githubusercontent.com/4742081/122794507-9d51cb00-d2bc-11eb-8867-c4e8e056f134.png)

You chose to hide intermediate collections, which is arguably good from an end-user's perspective, but then you need to take care that the names of the remaining datasets confer usable information. In the screenshot above, for example, it would be extremely tedious to track the provenance of that collection 331. It may be better to just call it "Annotated variants" then.

Personally I'd remove some of your WF step labels (not step output labels) and use them as the output names. WF step labels make it harder, imo, to see what a workflow does cause you have to click all the boxes in the editor to know which tool is getting used there (just my own personal opinion though).

wm75 commented:
and finally a question: why are you using an ivar variants AF threshold of 0.7 for calling variants, but a consensus threshold of 0.5. Wouldn't that occassionally incorporate a variant base into your consensus that isn't even reported in the annotated vcf?

I would have expected equal values here or, if anything, a *higher* consensus threshold, i.e., call and annotate relatively aggressively, but only build the consensus based on what looks credible?

mvdbeek commented:
> Personally I'd remove some of your WF step labels (not step output labels) and use them as the output names. WF step labels make it harder, imo, to see what a workflow does cause you have to click all the boxes in the editor to know which tool is getting used there (just my own personal opinion though).

That's a shortcoming of the editor that is relatively easy to fix. If the labels communicate the intent of the step we should keep it, but that is a very good point, we need to fix that in the editor. It's particularly painful for labelled subworkflows.

bwlang commented:
I've done some reworking of this so it can be used as a subworkflow by a top-level ONT and top-level Illumina workflow.
@pvanheus : does that appeal to you? should I PR that into your fork?

pvanheus commented:
Thanks @mvdbeek @wm75 @bgruening @drosofff I have updated the workflow and incorporated your suggestions. So:
1. The latest versions of tools are used
2. The reference consensus is renamed at the start of the workflow (and the rename is made more strict to only apply to this one case)
3. The pangolin and nextclade are run on concatenated FASTA
4. Each collection tool output does not use `on #{input}` in its labeling.

pvanheus commented:
btw this workflow works when tested against https://galaxy.sanbi.ac.za

wm75 commented on workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis-test.yml:
```diff
@@ -0,0 +1,35 @@
+      filetype: bed
+    Amplicon Primer Info:
+      class: File
+      location: 'https://covid19.galaxyproject.org/artic/ARTIC_SARS_CoV-2_amplicon_info_v3.tsv'
```
This file is needed for the -f option of ivar trim. You should probably make use of this option also in the training material, add the file to zenodo record 4944064, and get all test data from the same source then.

mvdbeek replied:
@pvanheus I think that comment still applies ?

pvanheus replied:
The Amplicon Primer Info file is now computed (by the `ivar trim` tool) so doesn't need to be provided by the user.

wm75 commented:
cosmetic change: The WF annotation is
`"annotation": "Find &amp; annotate variants in SARS-CoV-2 samples sequenced with Illumina ARTICv3",`
currently. Don't know how the `&amp;` slipped in there. Maybe the WF editor doesn't like `&`?

wm75 commented on workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis.ga:
```diff
@@ -0,0 +1,1142 @@
+        },
+        "19": {
+            "annotation": "",
+            "content_id": "toolshed.g2.bx.psu.edu/repos/bgruening/text_processing/tp_cat/0.1.1",
```
Just out of interest: @mvdbeek is this tool or "Collapse collection" better suited for concatenating the elements of a collection?

mvdbeek replied:
tp_cat is simpler and has fewer options. It's a thin wrapper around cat, while `Collapse collection` has options to add the element_identifier and to keep one header line. Collapse collection` would probably reach a limit in the length of the generated command line for very big collections (https://github.com/bgruening/galaxytools/pull/1135) ... which could be fixed of course.

So for this use tp_cat is probably better.

pvanheus commented:
> I've done some reworking of this so it can be used as a subworkflow by a top-level ONT and top-level Illumina workflow.
> @pvanheus : does that appeal to you? should I PR that into your fork?
> Also addressed Wolfgang's pangolin concern.

@bwlang said on chat that he will do his PR as an update after this one is merged.

bwlang commented:
Any reason not to merge now?

mvdbeek commented on workflows/sars-cov-2-variant-calling/sars-cov-2-pe-illumina-artic-ivar-analysis/pe-wgs-ivar-analysis.ga:
```diff
@@ -0,0 +1,1142 @@
+            "name": "Peter van Heusden"
+        }
+    ],
+    "format-version": "0.1",
```
```suggestion
    "release": "0.1",
    "format-version": "0.1",
```

wm75 reviewed (approved):
Looks like a really mature first release version to me!
Congrats and thanks @pvanheus!

drosofff commented:
Thank you @pvanheus ! If I did not misunderstood, I really really like the no-need-for-primer-bed feature !
